Library/Homebrew/livecheck/strategy/github_release.rb

```diff
@@ -0,0 +1,136 @@
+
+        NICE_NAME = "GitHub - Releases"
+
+        # A priority of zero causes livecheck to skip the strategy. We do this
+        # for {GithubLatest} so we can selectively apply the strategy using
+        # `strategy :github_latest` in a `livecheck` block.
+        PRIORITY = 9
+
+        # The `Regexp` used to determine if the strategy applies to the URL.
+        URL_MATCH_REGEX = %r{
+          ^https?://github\.com
+          /(?:downloads/)?(?<username>[^/]+) # The GitHub username
+          /(?<repository>[^/]+)              # The GitHub repository name
+        }ix.freeze
+
+        # The default regex used to identify a version from a tag when a regex
+        # isn't provided.
+        DEFAULT_REGEX = /v?(\d+(?:\.\d+)+)/i.freeze
+
+        # Keys in the JSON that could contain the version.
+        VERSION_KEYS = ["name", "tag_name"].freeze
+
+        # Whether the strategy can be applied to the provided URL.
+        #
+        # @param url [String] the URL to match against
+        # @return [Boolean]
+        sig { params(url: String).returns(T::Boolean) }
+        def self.match?(url)
+          URL_MATCH_REGEX.match?(url)
+        end
+
+        # Uses the regex to match text in the content or, if a block is
+        # provided, passes the page content to the block to handle matching.
+        # With either approach, an array of unique matches is returned.
+        #
+        # @param content [String] the page content to check
+        # @param regex [Regexp, nil] a regex used for matching versions in the
+        #   content
+        # @return [Array]
+        sig {
+          params(
+            content: T::Array[T::Hash[String, T.untyped]],
+            regex:   T.nilable(Regexp),
+            block:   T.nilable(Proc),
+          ).returns(T::Array[String])
+        }
+        def self.versions_from_content(content, regex, &block)
+          if block
+            block_return_value = regex.present? ? yield(content, regex) : yield(content)
+            return Strategy.handle_block_return(block_return_value)
+          end
+
+          values = []
+          content.each do |release|
+            next if release.blank?
+
+            VERSION_KEYS.each do |key|
+              match = release[key].match(regex)
+
+              next if match.blank?
+
+              values.push(match.to_s)
+              break
+            end
+          end
+
+          values.compact.uniq
+        end
+
+        # Generates a URL and regex (if one isn't provided) and passes them
+        # to {PageMatch.find_versions} to identify versions in the content.
+        #
+        # @param url [String] the URL of the content to check
+        # @param regex [Regexp] a regex used for matching versions in content
+        # @return [Hash]
+        sig {
+          params(
+            url:     String,
+            regex:   T.nilable(Regexp),
+            _unused: T.nilable(T::Hash[Symbol, T.untyped]),
+            block:   T.nilable(Proc),
```

Comment:
Do these have to all be nilable?

Reply:
Yes. For example, `regex` is only present if a formula/cask has a `livecheck` block using `#regex`. It's not present if a formula/cask doesn't contain a `livecheck` block or has one that doesn't use `#regex`.

Reply:
When [`#find_versions` is called in `Livecheck#latest_version`](https://github.com/Homebrew/brew/blob/22fb173e396fd02f32334a4d1b2c1cbbc669ae96/Library/Homebrew/livecheck/livecheck.rb#L740), the `regex` value from the formula/cask's `livecheck` object is passed to the method. [The `Livecheck` DSL initializes `@regex` to `nil`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/livecheck.rb#L26), so we pass a `nil` value to `#find_versions` when a regex isn't provided in a `livecheck` block. In that scenario, the `regex` argument is `nil` (i.e., the default value only applies when the argument isn't present). Unless I'm mistaken, that default value will never apply due to how we call `#find_versions`.

Reply:
> If you take the suggestion below it looks like regex doesn't need to be nilable

Yes. Having a default `nil` regex and then setting all nil values to a default later makes little sense to me.

I see `T.nilable` used all the time with Sorbet which removes one of the main benefits of the type system that handles `nil` properly: not needing to constantly be checking for `nil` values.